includes/class-newspack-popups-donations.php

```diff
@@ -37,122 +37,156 @@ public static function instance() {
+			(int) \get_option( \Newspack\Donations::DONATION_PRODUCT_ID_OPTION, 0 ) :
+			0;
+		$newspack_donation_product    = $newspack_donation_product_id ? \wc_get_product( $newspack_donation_product_id ) : null;
+		$newspack_child_products      = $newspack_donation_product ? $newspack_donation_product->get_children() : [];
+
+		/**
+		 * Allows other plugins to designate additional WooCommerce products by ID that should be considered donations.
+		 *
+		 * @param int[] $product_ids Array of WooCommerce product IDs.
+		 */
+		$other_donation_products = \apply_filters( 'newspack_popups_donation_products', [] );
+		$all_donation_products   = array_values( array_merge( $newspack_child_products, $other_donation_products ) );
+		$orders_data             = [];
+
+		foreach ( $orders as $order ) {
+			$order_data  = $order->get_data();
+			$order_items = array_map(
+				function( $item ) {
+					return $item->get_product_id();
+				},
+				array_values( $order->get_items() )
+			);
+
+			// Only count orders that include donation products as donations.
+			if ( 0 < count( array_intersect( $order_items, $all_donation_products ) ) ) {
+				$orders_data[] = [
+					'order_id' => $order_data['id'],
+					'date'     => date_format( date_create( $order_data['date_created'] ), 'Y-m-d' ),
+					'amount'   => $order_data['total'],
+				];
+			}
 		}
 
-		// If there was a donation, update the client data.
-		Newspack_Popups_Segmentation::update_client_data(
-			get_post_meta( $wc_order_data['id'], Newspack_Popups_Segmentation::NEWSPACK_SEGMENTATION_CID_NAME, true ),
-			[
-				'donation' => [
-					'order_id' => $wc_order_data['id'],
-					'date'     => date_format( date_create( $wc_order_data['date_created'] ), 'Y-m-d' ),
-					'amount'   => $wc_order_data['total'],
-				],
-			]
-		);
+		return $orders_data;
 	}
 
 	/**
-	 * Update WC order with the client id from hidden form field.
+	 * When a new WooCommerce order is created with the client ID meta,
+	 * log a new donation event to that client ID.
 	 *
-	 * @param String $order_id WC order id.
+	 * @param int         $order_id Order ID.
+	 * @param WC_Order    $order Order object.
+	 * @param string|null $client_id Client ID to associate with the donation. If not passed, will attempt to get from the current session.
 	 */
-	public static function woocommerce_checkout_update_order_meta( $order_id ) {
-		if ( ! empty( $_POST[ Newspack_Popups_Segmentation::NEWSPACK_SEGMENTATION_CID_NAME ] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Missing
-			update_post_meta( $order_id, Newspack_Popups_Segmentation::NEWSPACK_SEGMENTATION_CID_NAME, sanitize_text_field( $_POST[ Newspack_Popups_Segmentation::NEWSPACK_SEGMENTATION_CID_NAME ] ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
+	public static function create_donation_event( $order_id, $order, $client_id = null ) {
+		if ( ! $client_id ) {
+			$client_id = \Newspack_Popups_Segmentation::get_client_id();
+		}
+
+		if ( $client_id ) {
```

Comment:
NIT: this could return early for readability

Reply:
Good idea, done in c009c20